File: src/main/webapp/common.templates/java/dao/orm/hibernate/__classes__DAO.java
#header()
package ${aib.getRootPackageName(true)}.#getDAOPackageName();

import java.util.*;
import java.util.logging.Logger;

import java.sql.*;

import org.hibernate.*;
import org.hibernate.cfg.*;

import ${aib.getRootPackageName()}.exception.ProcessingException;

#set( $imports = [ "#getPrimaryKeyPackageName()", "#getBOPackageName()" ] )
#importStatements( $imports )

import ${aib.getRootPackageName()}.#getDAOPackageName().*;

/** 
 * Implements the Hibernate persistence processing for business entity ${classObject.getName()}.
 *
 * @author $aib.getAuthor()
 */

// AIB : \#getDAOClassDecl()
#getDAOClassDecl()
// ~AIB
{
    /**
     * default constructor
     */
    public ${classObject.getName()}DAO()
    {
    }

    /**
     * Retrieves a ${classObject.getName()} from the persistent store, using the provided primary key. 
     * If no match is found, a null ${classObject.getName()} is returned.
     * <p>
     * @param       pk
     * @return      ${classObject.getName()}
     * @exception   ProcessingException
     */
    public ${classObject.getName()} find${classObject.getName()}( ${classObject.getName()}PrimaryKey pk ) 
    throws ProcessingException
    {
        if (pk == null)
        {
            throw new ProcessingException("${classObject.getName()}DAO.find${classObject.getName()}(...) cannot have a null primary key argument");
        }
    
		Query query 	= null;		
		${classObject.getName()} businessObject = null;
		
	    StringBuilder fromClause = new StringBuilder( "from ${aib.getRootPackageName(true)}.#getBOPackageName().${classObject.getName()} as ${classObject.getName().toLowerCase()} where " );
		
        Session session = null;
        Transaction tx = null;

        try
        {
            session = currentSession();
            tx      = currentTransaction(session);
            
// AIB : \#getHibernateFindFromClause()
#getHibernateFindFromClause()
// ~AIB

	 		query = session.createQuery( fromClause.toString() );
	 		
	 		if ( query != null )
	 		{
                businessObject = new ${classObject.getName()}();
                businessObject.copy((${classObject.getName()})query.list().iterator().next());
	 		}

			commitTransaction(tx);						
		}
		catch( Throwable exc )
		{
			businessObject = null;
			exc.printStackTrace();
			throw new ProcessingException( "${classObject.getName()}DAO.find${classObject.getName()} failed for primary key " + pk + " - " + exc );		
		}		
		finally
		{
			try
			{
				closeSession();			
			}
			catch( Throwable exc )
			{		
				LOGGER.info( "${classObject.getName()}DAO.find${classObject.getName()} - Hibernate failed to close the Session - " + exc );
			}
		}		    
				
        return( businessObject );
    }
        
##if ( $aib.hasIdentity( $classObject ) == true )		
	    /**
	     * returns a Collection of all ${classObject.getName()}s
	     * @return		ArrayList<${classObject.getName()}>
	     * @exception   ProcessingException
	     */
	    public ArrayList<${classObject.getName()}> findAll${classObject.getName()}()
	    throws ProcessingException
	    {
			ArrayList<${classObject.getName()}> list = new ArrayList<${classObject.getName()}>();
			ArrayList<${classObject.getName()}> refList = new ArrayList<${classObject.getName()}>();
			Query query 							= null;		
			StringBuilder buf 						= new StringBuilder( "from ${aib.getRootPackageName(true)}.#getBOPackageName().${classObject.getName()}" );

			try
			{
				Session session = currentSession();

		 		query = session.createQuery( buf.toString() );
		 		
				if ( query != null )
				{
					list = (ArrayList<${classObject.getName()}>)query.list();
					${classObject.getName()} tmp = null;
					
	                for (${classObject.getName()} listEntry : list)
	                {
	                    tmp = new ${classObject.getName()}();
	                    tmp.copyShallow(listEntry);
	                    refList.add(tmp);
	                }
				}
			}
			catch( Throwable exc )
			{
				exc.printStackTrace();		
				throw new ProcessingException( "${classObject.getName()}DAO.findAll${classObject.getName()} failed - " + exc );		
			}		
			finally
			{
				try
				{
					closeSession();			
				}
				catch( Throwable exc )
				{		
					LOGGER.info( "${classObject.getName()}DAO.findAll${classObject.getName()} - Hibernate failed to close the Session - " + exc );
				}		
			}
			
			if ( list.size() <= 0 )
			{
				LOGGER.info( "${classObject.getName()}DAO:findAll${classObject.getName()}s() - List is empty.");
			}
	        
			return( refList );		        
	    }
		

	    /**
	     * Inserts a new ${classObject.Name} into the persistent store.
	     * @param       businessObject
	     * @return      newly persisted ${classObject.getName()}
	     * @exception   ProcessingException
	     */
	    public ${classObject.getName()} create${classObject.getName()}( ${classObject.getName()} businessObject )
	    throws ProcessingException
	    {
		    Transaction tx 	= null;
			Session session	= null;
			
	    	try
	    	{    				
				session	= currentSession();
				tx 		= currentTransaction( session );
		
				session.save( businessObject );
				commitTransaction( tx );	
				
			}
			catch( Throwable exc )
			{
				try
				{
					if ( tx != null )
						rollbackTransaction( tx );				
				}
				catch( Throwable exc1 )
				{
					LOGGER.info( "${classObject.getName()}DAO.create${classObject.getName()} - Hibernate failed to rollback - " + exc1 );
				}
				exc.printStackTrace();			
				throw new ProcessingException( "${classObject.getName()}DAO.create${classObject.getName()} failed - " + exc );
			}		
			finally
			{
				try
				{
					session.flush();
					closeSession();			
				}
				catch( Throwable exc )
				{		
					LOGGER.info( "${classObject.getName()}DAO.create${classObject.getName()} - Hibernate failed to close the Session - " + exc );
				}
			}		
			
	        // return the businessObject
	        return(  businessObject );
	    }
	    
    /**
     * Stores the provided ${classObject.getName()} to the persistent store.
     *
     * @param       businessObject
     * @return      ${classObject.getName()}	stored entity
     * @exception   ProcessingException 
     */
    public ${classObject.getName()} save${classObject.getName()}( ${classObject.getName()} businessObject )
    throws ProcessingException
    {
		Transaction tx 	= null;
		Session session = null;
		
    	try
    	{
			session = currentSession();
			tx		= currentTransaction( session );
	
			session.update( businessObject );
			commitTransaction( tx );	
		}
		catch( Throwable exc )
		{
			try
			{
				if ( tx != null )
					rollbackTransaction( tx );
			}
			catch( Throwable exc1 )
			{
				LOGGER.info( "${classObject.getName()}DAO.save${classObject.getName()} - Hibernate failed to rollback - " + exc1 );
			}
			exc.printStackTrace();			
			throw new ProcessingException( "${classObject.getName()}DAO.save${classObject.getName()} failed - " + exc );		
		}		
		finally
		{
			try
			{
				session.flush();
				closeSession();			
			}
			catch( Throwable exc )
			{		
				LOGGER.info( "${classObject.getName()}DAO.save${classObject.getName()} - Hibernate failed to close the Session - " + exc );
			}
		}		    
        return( businessObject );
    }
    
    /**
    * Removes a ${classObject.getName()} from the persistent store.
    *
    * @param        pk		identity of object to remove
    * @exception    ProcessingException
    */
    public void delete${classObject.getName()}( ${classObject.getName()}PrimaryKey pk ) 
    throws ProcessingException 
    {
	    Transaction tx 	= null;
	    Session session = null;
	    
    	try
    	{    	
			${classObject.getName()} bo = find${classObject.getName()}(pk);    	
			
			session = currentSession();
			tx		= currentTransaction( session );						
			session.delete( bo );
			commitTransaction( tx );
		}
		catch( Throwable exc )
		{
			try
			{
				if ( tx != null )
					rollbackTransaction( tx );				
			}
			catch( Throwable exc1 )
			{
				LOGGER.info( "${classObject.getName()}DAO.delete${classObject.getName()} - Hibernate failed to rollback - " + exc1 );
			}
			exc.printStackTrace();			
			throw new ProcessingException( "${classObject.getName()}DAO.delete${classObject.getName()} failed - " + exc );					
		}		
		finally
		{
			try
			{
				session.flush();
				closeSession();			
			}
			catch( Throwable exc )
			{		
				LOGGER.info( "${classObject.getName()}DAO.delete${classObject.getName()} - Hibernate failed to close the Session - " + exc );
			}
		}
    }

##end ##if ( $aib.hasIdentity( $classObject ) == true )		

// AIB : \#outputDAOFindAllImplementations()
##outputDAOFindAllImplementations()
// ~AIB


//*****************************************************
// Attributes
//*****************************************************
	private static final Logger LOGGER 	= Logger.getLogger(${classObject.getName()}.class.getName());

}

File: tech.stack.packages/ASP.NET Core/persistence/dao/orm/hibernate/__classes__DAO.cs
#header()
#set( $appName = $aib.getApplicationNameFormatted() )
#set( $className = $classObject.getName() )
using System;
using System.Collections;
using System.Collections.Generic;
using Microsoft.Extensions.Logging;
using NHibernate;
using NHibernate.Cfg;

using ${appName}.Exceptions;
using ${appName}.Models;
using ${appName}.PrimaryKeys;
using ${appName}.Persistence;

namespace ${appName}.DAOs
{
    /// <summary>
    /// Responsible for interacting with the ORM abstraction to create, read, update, 
    /// and delete an ${appName} entity along with its associated entities
    /// </summary>
	public class ${className}DAO
	{ 
        ///<summary>
        /// default constructor, using dependency injection to acquire an ILogger<${className}DAO> interface
        /// <param name="_logger|></para>
        /// </summary>
		public ${className}DAO( ILogger<${className}DAO> _logger )
		{
			logger = _logger;
		}
		
        ///<summary>
        /// Retrieves a ${className} from the persistent store, using the provided primary key. 
        /// If no match is found, a null ${className} is returned.
        /// <paramm name="pk"></para>
        /// </summary>

    	public ${className} find${className}( ${className}PrimaryKey pk ) 
    	{
    		$className model = null;
    		
        	if (pk == null)
        	{
            	throw( new ProcessingException("${className}DAO.find${className}(...) cannot have a null primary key argument") );
        	}

			using (ISession session = FrameworkPersistenceHelper.OpenSession())
			{
	            using (ITransaction transaction = session.BeginTransaction())
    	        {
		        	try
	    	    	{
	                    model = new ${classObject.getName()}();
	                    model.copy( session.Get<${className}>(pk.getFirstKey()) );
					}
					catch( Exception exc )
					{
						model = null;
						Console.WriteLine("Exception caught: {0}", exc);
						throw ( new ProcessingException( "${className}DAO.find${className} failed for primary key " + pk + " - " + exc.ToString() ) );		
					}		
					finally
					{
					}
				}
			}		    
					
        	return( model );
    	}
	    
#if ( $aib.hasIdentity( $classObject ) == true )		
        ///<summary>
        /// returns a List of all the ${className} entities
        /// <returns></returns>
        ///</summary>
	    public List<${className}> findAll${className}()
	    {
			List<${className}> refList = new List<${className}>();
			IList list;

     		using (ISession session = FrameworkPersistenceHelper.OpenSession())
     		{
	            using (ITransaction transaction = session.BeginTransaction())
    	        {
					try
					{
						string buf 	= "from ${className}";
				 		IQuery query = session.CreateQuery( buf );
				 		
						if ( query != null )
						{
							list = query.List();
							${className} model = null;
							
			                foreach (${className} listEntry in list)
			                {
			                    model = new ${className}();
			                    model.copyShallow(listEntry);
			                    refList.Add(model);
			                }
						}
					}
					catch( Exception exc )
					{
						Console.WriteLine("Exception caught: {0}", exc);
						throw ( new ProcessingException( "${className}DAO.findAll${className} failed - " + exc.ToString() ) );		
					}		
					finally
					{
					}
				}
			}
			if ( refList.Count <= 0 )
			{
				logger.LogInformation( "${className}DAO:findAll${className}s() - List is empty.");
			}
	        
			return( refList );		        
	    }
		
        ///<summary>
        /// Inserts a new ${classObject.Name} model into the persistent store and returns 
        /// <param name="model"></para>
        /// <returns></returns>
        ///</summary>
	    public ${className} create${className}( ${className} model )
	    {
     		using (ISession session = FrameworkPersistenceHelper.OpenSession())
     		{
	            using (ITransaction transaction = session.BeginTransaction())
    	        {
			    	try
			    	{
			    		if ( model != null )
			    		{
		    	        	session.Save(model);
		        	    	transaction.Commit();
		        	    }
		        	    else
						{
							string errMsg = "${className}DAO.create${className} - null model provided but not allowed";
							logger.LogInformation( errMsg );
							throw ( new ProcessingException( errMsg ) );		
						}		        	    
					}
					catch( Exception exc )
					{
						string errMsg = "${className}DAO.create${className} - Hibernate failed to rollback - " + exc.ToString();
						Console.WriteLine("Exception caught: {0}", exc.ToString());
						logger.LogInformation( errMsg );
						throw ( new ProcessingException( errMsg ) );		
					}		
					finally
					{
					}		    
			        return( model );
				}
			}	 
	    }
		    
        ///<summary>
        /// Updates the provided ${className} model to the persistent store.
        /// <param name="model"></para>
        /// <returns></returns>
        ///</summary>
	    public ${className} save${className}( ${className} model )
	    {
     		using (ISession session = FrameworkPersistenceHelper.OpenSession())
     		{
	            using (ITransaction transaction = session.BeginTransaction())
    	        {
			    	try
			    	{
		    	        session.Update(model);
		        	    transaction.Commit();
					}
					catch( Exception exc )
					{
						string errMsg = "${className}DAO.save${className} - Hibernate failed to rollback - " + exc.ToString();
						Console.WriteLine("Exception caught: {0}", exc.ToString());
						logger.LogInformation( errMsg );
						throw ( new ProcessingException( errMsg ) );		
					}		
					finally
					{
					}		    
			        return( model );
				}
			}	 
	    }
	    
        ///<summary>
        /// Removes the associated ${className} model from the persistent store.
        /// <param name="pk"></para>
        /// <returns></returns>
        ///</summary>
	    public bool delete${className}( ${className}PrimaryKey pk ) 
	    {
	    	bool deleted = false;
	    	
			using (ISession session = FrameworkPersistenceHelper.OpenSession())
			{
                using (ITransaction transaction = session.BeginTransaction())
                {
			    	try
	    			{
	    				${className} model = find${className}(pk);    	
	    			
	    				session.Delete( model );
                    	transaction.Commit();
                    	deleted = true;
                	}	    
					catch( Exception exc )
					{
						Console.WriteLine("Exception caught: {0}", exc );
						logger.LogInformation( "${className}DAO.delete${className} failed - " + exc.ToString() );
						throw ( new ProcessingException( "${className}DAO.delete${className} failed - " + exc.ToString() ) );					
					}		
					finally
					{
					}
				}	    			    	
			}
			
			return deleted;
	    }

#end ##if ( $aib.hasIdentity( $classObject ) == true )		

//*****************************************************
// Attributes
//*****************************************************
		private readonly ILogger<${className}DAO> logger;
	}

}


